feat: typescript 3 support

alan-agius4 opened the pull request:
This adds support for TypeScript 3.0.

Notes: 
- for the time being `tsickle` doesn't support `TypeScript 3.0`
- 3 unit tests from `program_spec.ts` have been `xit`. Anyone willing to help. Please feel free.

Closes: #25200

//cc @mhevery

jotatoledo commented on integration/typings_test_ts30/package.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "angular-integration",
+  "description": "Assert that users with TypeScript 2.9 can type-check an Angular application",
```
TypeScript 3.0?

alan-agius4 committed:
feat: add TypeScript 3 support

alan-agius4 commented:
I confirm, I sign it.

Thanks Igor

IgorMinar commented:
Please ignore the cla bot. I rebased and updated your pr. Still looking at
at failing tests.

On Thu, Aug 9, 2018, 9:14 PM Alan Agius <notifications@github.com> wrote:

> I confirm, I sign it.
>
> Thanks Igor
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/25275#issuecomment-411970154>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANM6OcLyKNd4paN9jhMxBgDLsvC5ZdWks5uPQiLgaJpZM4VtgO->
> .
>

JoostK commented:
@alan-agius4 they still seem to fail in Travis

IgorMinar commented:
@alan-agius4 still looking into it.

mary-poppins commented:
You can preview 743a985 at https://pr25275-743a985.ngbuilds.io/.
You can preview d47e023 at https://pr25275-d47e023.ngbuilds.io/.
You can preview 2a37db7 at https://pr25275-2a37db7.ngbuilds.io/.
You can preview 5fae320 at https://pr25275-5fae320.ngbuilds.io/.
You can preview f87fbd8 at https://pr25275-f87fbd8.ngbuilds.io/.
You can preview 9019299 at https://pr25275-9019299.ngbuilds.io/.
You can preview 7b674fe at https://pr25275-7b674fe.ngbuilds.io/.
You can preview f8c2d39 at https://pr25275-f8c2d39.ngbuilds.io/.
You can preview 8bf9ea8 at https://pr25275-8bf9ea8.ngbuilds.io/.
You can preview 3ebd8fc at https://pr25275-3ebd8fc.ngbuilds.io/.
You can preview 4cdd353 at https://pr25275-4cdd353.ngbuilds.io/.
You can preview f575cad at https://pr25275-f575cad.ngbuilds.io/.
You can preview 2874da4 at https://pr25275-2874da4.ngbuilds.io/.
You can preview 8c2f485 at https://pr25275-8c2f485.ngbuilds.io/.
You can preview 09faa8d at https://pr25275-09faa8d.ngbuilds.io/.
You can preview b9427a0 at https://pr25275-b9427a0.ngbuilds.io/.
You can preview fc59117 at https://pr25275-fc59117.ngbuilds.io/.
You can preview 73a8f93 at https://pr25275-73a8f93.ngbuilds.io/.
You can preview 6e27660 at https://pr25275-6e27660.ngbuilds.io/.
You can preview a612d0c at https://pr25275-a612d0c.ngbuilds.io/.
You can preview 438dea7 at https://pr25275-438dea7.ngbuilds.io/.
You can preview f725b36 at https://pr25275-f725b36.ngbuilds.io/.
You can preview 5197c21 at https://pr25275-5197c21.ngbuilds.io/.
You can preview b2ee232 at https://pr25275-b2ee232.ngbuilds.io/.
You can preview baaf770 at https://pr25275-baaf770.ngbuilds.io/.
You can preview 330418e at https://pr25275-330418e.ngbuilds.io/.
You can preview 9c1f99f at https://pr25275-9c1f99f.ngbuilds.io/.

alxhub committed:
test(compiler-cli): fix the incremental ngc tests so that they run under bazel

IgorMinar committed:
test(compiler-cli): improve testing harness for incremental compilation

In tsc 3.0 the check that enables program structure reuse in tryReuseStructureFromOldProgram has changed
and now uses identity comparison on arrays within CompilerOptions. Since we recreate the options
on each incremental compilation, we now fail this check.

After this change the default set of options is reused in between incremental compilations, but we still
allow options to be overriden if needed.

mary-poppins commented:
You can preview b58e519 at https://pr25275-b58e519.ngbuilds.io/.

mary-poppins commented:
You can preview 778b097 at https://pr25275-778b097.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/transformers/compiler_host.ts:
```diff
@@ -393,6 +393,9 @@ export class TsCompilerAotCompilerTypeCheckHostAdapter implements ts.CompilerHos
   getSourceFile(
       fileName: string, languageVersion: ts.ScriptTarget,
       onError?: ((message: string) => void)|undefined): ts.SourceFile {
+        if (fileName.endsWith('@angular/core/src/di/injection_token.d.ts')) {
```
Debugging code; remove.

alxhub commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -59,7 +59,7 @@ describe('ng program', () => {
   function compile(
       oldProgram?: ng.Program, overrideOptions?: ng.CompilerOptions, rootNames?: string[],
-      host?: CompilerHost): {program: ng.Program, emitResult: ts.EmitResult} {
+      host?: CompilerHost): {program: ng.Program, emitResult: ts.EmitResult, host: ng.CompilerHost} {
```
No need to return the host, it's not used.

alxhub commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -75,7 +75,30 @@ describe('ng program', () => {
+    const originalGetSourceFile = host.getSourceFile;
+    const cache = new Map<string, ts.SourceFile>();
+    host.getSourceFile = function(fileName: string): ts.SourceFile {
+      if (fileName.endsWith('@angular/core/src/di/injection_token.d.ts')) {
```
Debugging code; remove.

alxhub commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -75,7 +75,30 @@ describe('ng program', () => {
+        debugger;
+      }
+      const sf = originalGetSourceFile.call(host, fileName) as ts.SourceFile;
+      if (sf && cache.has(sf.fileName)) {
```
Maybe put the whole section in an `if (sf)` block.

alxhub commented on packages/compiler-cli/test/transformers/program_spec.ts:
```diff
@@ -267,18 +290,22 @@ describe('ng program', () => {
-        compile(p2);
+        const p1 = compile(undefined, undefined, undefined, host).program;
+        const p2 = compile(p1, undefined, undefined, host).program;
+        debugger;
```
Remove.

IgorMinar commented:
Oops.. @alxhub I just squashed your commits but forgot to clean them up. I'll do that tomorrow and get this PR merged. Thanks for your the help today.

IgorMinar commented:
caretaker note: [presubmit results](https://test.corp.google.com/ui#id=OCL:208075357:BASE:208909396:1534377865110:d6fbb63) are green I just can't push the status right now.

mary-poppins commented:
You can preview 6c6be72 at https://pr25275-6c6be72.ngbuilds.io/.

mary-poppins commented:
You can preview 163b04e at https://pr25275-163b04e.ngbuilds.io/.

matsko commented:
@alan-agius4 please rebase

alan-agius4 commented:
@matsko, rebase done.

mary-poppins commented:
You can preview 083bb3e at https://pr25275-083bb3e.ngbuilds.io/.

alan-agius4 commented:
Can someone try to re-trigger the failed workflows? Thanks

On Fri, 24 Aug 2018 at 07:17, Mary Poppins <notifications@github.com> wrote:

> You can preview 083bb3e
> <https://github.com/angular/angular/commit/083bb3e5d226b5d23e92aefa99e7ea2583f94518>
> at https://pr25275-083bb3e.ngbuilds.io/.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/25275#issuecomment-415653575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQv-WoRBgoGJkXRv9n2Bpq3zeJ2VBbAqks5uT4yCgaJpZM4VtgO->
> .
>

mary-poppins commented:
You can preview daf7271 at https://pr25275-daf7271.ngbuilds.io/.

mary-poppins commented:
You can preview 33f693b at https://pr25275-33f693b.ngbuilds.io/.

matsko commented:
@alan-agius4 can you please rebase and push again. Master on CircleCI should be ok now.

alan-agius4 commented:
@matsko I m trying too, but I am getting `Already up to date.`

IgorMinar commented:
caretaker note: [presubmit results](https://test.corp.google.com/ui#id=OCL:208075357:BASE:208909396:1534377865110:d6fbb63) are green I just can't push the status right now.

matsko commented:
Sorry @alan-agius4 it still needs one last rebase. It looks like the auto-squashing of the `fixup!` commits is failing. Can you please interactively rebase and do the squashing yourself in the interactive mode? This way we can merge this without the git autosquash bugging out.

Just do `git rebase HEAD~10` and combine the fixup commits into the main commits.

Thank you.

alan-agius4 commented:
@matsko, rebase with fixups done. 

Need to a googler to verify the CLA.

mary-poppins commented:
You can preview 463194a at https://pr25275-463194a.ngbuilds.io/.
